src/classes/BDI_DataImportBatchStatus_TDTM.cls

```diff
@@ -35,193 +35,273 @@
-        BatchDataImportStatusAggregate() {
-            openStatusCount = 0;
-            importedStatusCount = 0;
-            failedStatusCount = 0;
+     * A class for determining the status of a single Data Import Batch
+     */
+    private class BatchDataImportStatus {
+        Boolean hasOpenRow = false;
+        Boolean hasImportedRow = false;
+        Boolean hasFailedRow = false;
+
+        /**
+         * @param batch A data import batch with a single row sub-query on the NPSP_Data_Imports__r relationship
+         */
+        void readQueryResult(DataImportBatch__c batch) {
+            if (batch.NPSP_Data_Imports__r != null) {
+                for (DataImport__c dataImport : batch.NPSP_Data_Imports__r) {
+                    readRowStatus(dataImport);
+                }
+            }
         }
 
         /**
-         * @description update the status count based on the incoming di status
-        * @param String status
-        */
-        void updateCounts(String status) {
-            if (String.isBlank(status) || status == BDI_DataImport_API.bdiDryRunValidated
-                                    || status == BDI_DataImport_API.bdiDryRunError) {
-                openStatusCount++;
-            } else if (status == BDI_DataImport_API.bdiImported) {
-                importedStatusCount++;
-            } else if (status == BDI_DataImport_API.bdiFailed) {
-                failedStatusCount++;
+         * Read the status of a single data import row from the inner query on the batch,
+         * and set one of the batch status flags to indicate that the specified type of row was found.
+         * @param dataImport The data import row to get the status of.
+         */
+        void readRowStatus(DataImport__c dataImport) {
+            String status = dataImport.Status__c;
+
+            if (isOpen(status)) {
+                this.hasOpenRow = true;
+            }
+
+            if (isImported(status)) {
+                this.hasImportedRow = true;
+            }
+
+            if (isFailed(status)) {
+                this.hasFailedRow = true;
+            }
+        }
+
+        /**
+         * After ingesting all of the Data Import Batch rows, determine the final batch status.
+         * @return {String} One of 'Failed - Needs Review', 'Open', or 'Completed'
+         */
+        String getBatchStatus() {
+            // if the batch has one failed row, the batch is failed regardless of the state of the other rows
+            if (this.hasFailedRow) {
+                return BDI_DataImport_API.BATCH_STATUS_FAILED;
             }
+
+            // if all rows are imported, and there are no failed or open rows, the batch is completed
+            if (this.hasImportedRow && !this.hasFailedRow && !this.hasOpenRow) {
+                return BDI_DataImport_API.BATCH_STATUS_COMPLETED;
+            }
+
+            // if none of the above apply, the batch is considered open
+            return BDI_DataImport_API.BATCH_STATUS_OPEN;
```

Comment:
If we make the change above, here we could just return BDI_DataImport_API.BATCH_STATUS_COMPLETED.

Reply:
@vmartinov-sforg An empty batch is technically an open batch. I don't think we can change the default case to be that the batch is completed because empty batches would get marked as completed.  That said, I do have some simplifications here that should make this read better.